### PR TITLE
UHF stability analysis

### LINKED ISSUE #585: Stability analysis
**Describe the feature you'd like**
With GHF a part of GQCP, it would be nice to be able to perform a stability analysis on the Hartree-Fock methods using the orbital rotation Hessian. The respective stability matrices and stability conditions for the other HF methods could be added as well. 

**Describe what the current code offers in relation to what it lacks**
At the moment, the code does not offer this feature.

### LINKED ISSUE #789: UHF stability analysis
**Describe the feature you'd like**
Implement the stability checks for the UHF wave function model.

**Describe what the current code offers in relation to what it lacks**
The stability checks are not yet part of the code.

### PR DESCRIPTION
**Pull request by xdvriend**
**Short description**
As a final step in adding stability analysis to GQCP, I will tackle the UHF stability conditions in this PR. 

**Related issues**
closes #789 
closes #585 

**Additional context**
Tases are the steps that need to be completed:
- [x] Implement spin-conserved stability matrix calculations.
- [x] Implement spin-unconserved stability matrix calculations.
- [x] Implement `UHFStabilityMatrices` class.
- [x] Test the UHF stability analysis.

**Don't forget:**
- [x] if new files are created: update the Doxygen file, the collective gqcp.hpp header and the CMake files

### CODE REVIEW

File: gqcp/include/QCModel/HF/UHF.hpp
Comment (lelemmen) [inline gqcp/include/QCModel/HF/UHF.hpp:376]: Maybe use the following APIs?

```suggestion
        const auto orbital_space = this->orbitalSpace();
        const auto& orbital_space_sigma = orbital_space.component(sigma);
        const auto& orbital_space_sigma_bar = orbital_space.component(sigma_bar);
```

File: gqcp/include/QCModel/HF/UHF.hpp
Comment (lelemmen) [inline gqcp/include/QCModel/HF/UHF.hpp:360]: Could you clearly indicate that this Hamiltonian is expressed in the non-orthogonal, 'restricted' spin-orbital basis of the AOs?

Please do so for the previous methods you've implemented as well.

File: gqcp/include/QCModel/HF/UHF.hpp
Comment (lelemmen) [inline gqcp/include/QCModel/HF/UHF.hpp:394]: I've looked at the implementation of all of the components of the stability matrices, and it seems that you're always transforming the 'restricted AO spin-orbitals' to the UHF basis (= your mixed MO basis).

Instead of doing so, maybe I could suggest an alternative approach?
- Let the user set up a `USQHamiltonian` in the basis of the orthonormal UHF spin-orbitals, obtained from a converged UHF calculation.
- Work with that orthonormal USQHamiltonian instead of having to transform the integrals in these methods, because it encapsulates the (orthonormal) two-electron operators `g_aaaa`, `g_aabb`, `g_bbaa` and `g_bbbb`.

File: gqcp/include/QCModel/HF/UHF.hpp
Comment (lelemmen) [inline gqcp/include/QCModel/HF/UHF.hpp:635]: In this approach, you're re-calculating `this->orbitalSpace()` many times. Could you make a separate variable for it and access that one accordingly?

File: gqcp/include/QCModel/HF/UHF.hpp
Comment (lelemmen) [inline gqcp/include/QCModel/HF/UHF.hpp:967]: For this method, `occupiedOrbitalEnergies` and `virtualOrbitalEnergies`,  could you maybe add a method that return a `SpinResolved<MatrixX<Scalar>>`, that holds the alpha and beta values together, analogously to what `orbitalSpace()` does?

You don't have to create a new class. We have a special helpful type `SpinResolved<T>` which encapsulates an alpha-`T` and a `beta`-T.

File: gqcp/tests/QCMethod/HF/UHF/QCMethod_UHF_stability_test.cpp
Comment (lelemmen) [inline gqcp/tests/QCMethod/HF/UHF/QCMethod_UHF_stability_test.cpp:100]:
```suggestion
    // This method should be internally unstable.
```

**Review by lelemmen — COMMENTED** [review]
Wow, what an implementation. This must have asked for a lot of focus! Since I am not familiar with the formulas, could you double-check them one last time for index errors?

I've added a couple of suggestions that need addressing before we can merge this PR. It's not that the implementation is wrong, but maybe we can improve them. I've given some general comments on `calculateMixedSpinConservedAComponent`, but the remarks should be applicable to the other methods as well.

File: gqcp/include/QCModel/HF/StabilityMatrices/UHFStabilityMatrices.hpp
Comment (xdvriend) [reply gqcp/include/QCModel/HF/StabilityMatrices/UHFStabilityMatrices.hpp:148]: Yes. Both the A' and B' matrix are square matrices with axis dimensions `(occ_a * virt_a) + (occ_b * virt_b)` The components of the A' and B' matrices don't necessarily have the same dimensions, but the A' and B' matrices doe.

File: gqcp/include/QCModel/HF/StabilityMatrices/UHFStabilityMatrices.hpp
Comment (xdvriend) [reply gqcp/include/QCModel/HF/StabilityMatrices/UHFStabilityMatrices.hpp:177]: Yep, same comments as before. The dimensions of the total A" and B" matrices are the same. This is not the case for their individual components.

File: gqcp/include/QCModel/HF/UHF.hpp
Comment (xdvriend) [reply gqcp/include/QCModel/HF/UHF.hpp:376]: Yep that's better! I must have looked over that API in the `SpinResolvedBase`module.

File: gqcp/include/Mathematical/Representation/Tensor.hpp
Comment (xdvriend) [reply gqcp/include/Mathematical/Representation/Tensor.hpp:431]: Ah yes, I believe you're right. But Since I don't have to split up the implementation, this will probably be removed entirely, as the other case is more general.

File: gqcp/include/QCModel/HF/StabilityMatrices/GHFStabilityMatrices.hpp
Comment (xdvriend) [reply gqcp/include/QCModel/HF/StabilityMatrices/GHFStabilityMatrices.hpp]: Adjusted in all stability matrix documentation.

File: gqcp/include/QCModel/HF/StabilityMatrices/UHFStabilityMatrices.hpp
Comment (xdvriend) [reply gqcp/include/QCModel/HF/StabilityMatrices/UHFStabilityMatrices.hpp:198]: Added in all classes of `StabilityMatrices`.

File: gqcp/include/QCModel/HF/StabilityMatrices/UHFStabilityMatrices.hpp
Comment (xdvriend) [reply gqcp/include/QCModel/HF/StabilityMatrices/UHFStabilityMatrices.hpp]: Adjusted in all stability matrix classes.

File: gqcp/include/QCModel/HF/UHF.hpp
Comment (xdvriend) [reply gqcp/include/QCModel/HF/UHF.hpp]: Adjusted throughout UHF model.

File: gqcp/include/QCModel/HF/UHF.hpp
Comment (xdvriend) [reply gqcp/include/QCModel/HF/UHF.hpp:635]: done in all 4 API's that used this approach.

File: gqcp/include/QCModel/HF/UHF.hpp
Comment (xdvriend) [reply gqcp/include/QCModel/HF/UHF.hpp:360]: Adjusted in all `QCModel::HF` classes.

File: gqcp/include/QCModel/HF/UHF.hpp
Comment (xdvriend) [reply gqcp/include/QCModel/HF/UHF.hpp:394]: Wouldn't this make it less clear for the user? Now the stability analysis is completely analogous for all Hartree-Fock methods. If I change this, GHF and RHF will have the same way of performing a stability analysis, with practically the same API's, but UHF will require an extra step, i.e. creating a second hamiltonian. Additionally I believe it can be quite confusing for the user to suddenly have to work with a `USQHamiltonian` while all other calculations in UHF are done using the `RSQHamiltonian`.

Comment (lelemmen) [reply gqcp/include/QCModel/HF/UHF.hpp:394]: It was implied that the RHF and GHF stability analyses would also use the `RSQHamiltonian` and `GSQHamiltonian` _in the HF orbital bases_.

I don't agree on your comment that this API can be confusing: a method developer should be aware that stability analysis requires a Hamiltonian in an orthonormal basis and SCF calculations requires the Hamiltonian in the original AO basis.

File: gqcp/include/Mathematical/Representation/Tensor.hpp
Comment (xdvriend) [reply gqcp/include/Mathematical/Representation/Tensor.hpp]: I've looked into this and the only thing I seem to find is going through an `Eigen::Map`. However, Eigen doesn't order the elements in the vector correctly for the method to return a matrix in the `C-like` ordering. Am I missing another possibility to do this?

**Comment by xdvriend** [comment]
@lelemmen, I double checked the indices with my python implementation and with my notes where I transformed everything from physicist's notation to chemist's notation: Everything should match.

I resolved all but two of the conversations you opened. I've left comments on this that are still unresolved as to why I haven't resolved them (yet). A brief summary of both:

- I'm not sure which `Eigen::Tensor` API could do this, as `Eigen::Map` creates a wrongly ordered vector.
- I think using the `USQHamiltonian` would make things a lot less clear. The UHF stability analysis would no longer be analogous to the other two and it seems counterintuitive to have to use a `USQHamiltonian` in a UHF method, as everywhere else in UHF the `RSQHamiltonian` is used.

With this, I have requested another review. Let me know what you think!

File: gqcp/include/QCModel/HF/UHF.hpp
Comment (xdvriend) [reply gqcp/include/QCModel/HF/UHF.hpp:394]: I see your point. I just don't fully see why it is okay for RHF and GHF to implicitly transform their Hamiltonian to an orthonormal basis, but not for UHF. 

Wouldn't it make more sense to treat all methods equally? So, if it is the responsibility of the developer to know when to use a Hamiltonian in the orthonormal basis, wouldn't it be best to apply this change to RHF and GHF as well?

> It was implied that the RHF and GHF stability analyses would also use the RSQHamiltonian and GSQHamiltonian in the HF orbital bases.

Or is that exactly what you meant by this? 😅

Comment (lelemmen) [reply gqcp/include/QCModel/HF/UHF.hpp:394]: > Or is that exactly what you meant by this? 😅

Yep! So we're on the same line here.

**Comment by xdvriend** [comment]
- I have double checked the indices of the methods with my python implementation and the transformations of the Pople formula's from physicists notation to chemist's notation I worked out. Everything should be correct!

- I have also modified all stability methods to require a Hamiltonian in MO basis. Ideally I would like the method to check whether you are working in MO basis and throw an exception when you are not, but for now, it is explained in the documentation.

- The reshape method is tweaked slightly so it throws an exception when you give impossible matrix dimensions as arguments. It uses `vector.reserve()` now as well to set aside the required amount of data for the vector. 

With this, everything from the last review should be addressed. And I have re-requested a review on this PR.

File: gqcp/include/Mathematical/Representation/Tensor.hpp
Comment (lelemmen) [inline gqcp/include/Mathematical/Representation/Tensor.hpp:425]: Maybe we could add an API `Tensor::numberOfElements` instead of this manual calculation?

File: gqcp/include/Mathematical/Representation/Tensor.hpp
Comment (lelemmen) [inline gqcp/include/Mathematical/Representation/Tensor.hpp:452]: Could you document why `r * cols + c` is the element you'd expect at `M(r,c)`?

File: gqcp/include/Mathematical/Representation/Tensor.hpp
Comment (lelemmen) [inline gqcp/include/Mathematical/Representation/Tensor.hpp:470]:
```suggestion
    bool hasEqualDimensionsAs(const Self& other) const {
```

File: gqcp/include/Mathematical/Representation/Tensor.hpp
Comment (lelemmen) [inline gqcp/include/Mathematical/Representation/Tensor.hpp:530]:
```suggestion
};
```

File: gqcp/tests/QCMethod/HF/GHF/QCMethod_GHF_stability_test.cpp
Comment (lelemmen) [inline gqcp/tests/QCMethod/HF/GHF/QCMethod_GHF_stability_test.cpp:54]: I'd like to propose the following way to generate the generalized molecular Hamiltonian.

```suggestion
    const auto hamiltonian_generalized = sq_hamiltonian.transformed(ghf_parameters.expansion());
```

In this method, there's not really a need for also transforming the spinor basis as well (since you don't use any of its APIs).

Also, I'm not really fond of the undescriptive variable name `h_mo`. I'd rather you change it to something more descriptive, like `hamiltonian_ghf` or something else. The name `h_mo` could also mean the core Hamiltonian in the MO basis, which is why I don't like it.

These comments of course also apply to the RHF and UHF test files.

File: gqcp/include/QCModel/HF/UHF.hpp
Comment (lelemmen) [inline gqcp/include/QCModel/HF/UHF.hpp:386]:
```suggestion
        const auto& g_aabb = usq_hamiltonian.twoElectron().alphaBeta().parameters();
```

For extra clarity, I think it's better to keep the `aabb` subscript, so we don't get confused which two-electron integrals we're using.

This comment also applies to the other methods, where you use `g_aaaa` and `g_bbbb`

**Review by lelemmen — APPROVED** [review]
After addressing the minor comments I've made, this PR is good to go!

Really nice work on these HF stability PRs. The APIs and their implementations are really clean!

**Pull request merged**
